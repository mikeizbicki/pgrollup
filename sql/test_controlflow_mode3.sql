SET client_min_messages TO WARNING;

-- this test ensures that do_rollup works even when there's no data in the table

UPDATE pgrollup_settings SET value='manual' WHERE name='default_mode';

create table test (
    id serial primary key,
    name text,
    num int
);

create materialized view test_rollup1 as (
    select name,count(*)
    from test
    group by name
);


select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');
select do_rollup('test_rollup1');

select drop_rollup('test_rollup1');
drop table test cascade;
UPDATE pgrollup_settings SET value='trigger' WHERE name='default_mode'

