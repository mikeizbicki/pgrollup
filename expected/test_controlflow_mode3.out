SET client_min_messages TO WARNING;
-- this test ensures that do_rollup works even when there's no data in the table
UPDATE pgrollup_settings SET value='manual' WHERE name='default_mode';
create table test (
    id serial primary key,
    name text,
    num int
);
create materialized view test_rollup1 as (
    select name,count(*)
    from test
    group by name
);
select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

select do_rollup('test_rollup1');
        do_rollup        
-------------------------
 (test_rollup1,test,1,0)
(1 row)

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select do_rollup('test_rollup1');
        do_rollup         
--------------------------
 (test_rollup1,test,1,26)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,27,26)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,27,26)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,27,26)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,27,26)
(1 row)

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,27,52)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,53,52)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,53,52)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,53,52)
(1 row)

select do_rollup('test_rollup1');
         do_rollup         
---------------------------
 (test_rollup1,test,53,52)
(1 row)

select drop_rollup('test_rollup1');
 drop_rollup 
-------------
 
(1 row)

drop table test cascade;
UPDATE pgrollup_settings SET value='trigger' WHERE name='default_mode'
